lint/location.go

```diff
@@ -14,43 +14,97 @@
 
 package lint
 
-import "fmt"
+import (
+	dpb "github.com/golang/protobuf/protoc-gen-go/descriptor"
+	"github.com/jhump/protoreflect/desc"
+)
 
-// Location describes a location in a source code file.
+// SyntaxLocation returns the location of the syntax definition in a file
+// descriptor.
 //
-// Note: positions are one-based.
-type Location struct {
-	Start Position `json:"start_position" yaml:"start_position"`
-	End   Position `json:"end_position" yaml:"end_position"`
+// If the location can not be found (for example, because there is no syntax
+// statement), it returns beginning of file.
+func SyntaxLocation(f *desc.FileDescriptor) *dpb.SourceCodeInfo_Location {
+	return sourceInfoRegistry.sourceInfo(f).findLocation([]int32{12}) // syntax == 12
 }
 
-// IsValid checks if the location is constructed properly and
-// returns true if so.
-func (l Location) IsValid() bool {
-	return l.Start.IsValid() &&
-		l.End.IsValid() &&
-		(l.End.Line > l.Start.Line ||
-			l.End.Line == l.Start.Line && l.End.Column >= l.Start.Column)
+// PackageLocation returns the location of the package definition in a file
+// descriptor.
+//
+// If the location can not be found (for example, because there is no package
+// statement), it returns beginning of file.
+func PackageLocation(f *desc.FileDescriptor) *dpb.SourceCodeInfo_Location {
+	return sourceInfoRegistry.sourceInfo(f).findLocation([]int32{2}) // package == 2
 }
 
-// String returns the string representation.
-func (l Location) String() string {
-	return fmt.Sprintf("{start: %s, end: %s}", l.Start, l.End)
+// DescriptorNameLocation returns the precise location for a descriptor's name.
+func DescriptorNameLocation(d desc.Descriptor) *dpb.SourceCodeInfo_Location {
+	// All descriptors seem to have `string name = 1`, so this conveniently works.
+	path := append(d.GetSourceInfo().Path, 1)
+	return sourceInfoRegistry.sourceInfo(d.GetFile()).findLocation(path)
 }
 
-// Position describes a one-based position in a source code file.
-type Position struct {
-	Line   int `json:"line_number" yaml:"line_number"`
-	Column int `json:"column_number" yaml:"column_number"`
+type sourceInfo map[string]*dpb.SourceCodeInfo_Location
+
+// findLocation returns the best location it can find for the given path.
+//
+// If the requested path can not be found, it climbs the ancestry tree until
+// it finds one, and ultimately returns a location corresponding to the
+// beginning of the file if it can not find anything.
+func (si sourceInfo) findLocation(path []int32) *dpb.SourceCodeInfo_Location {
+	// Base case: If we have no path left, send back a fake location that
+	// corresponds to "beginning of file".
+	if len(path) == 0 {
+		return &dpb.SourceCodeInfo_Location{
+			Span: []int32{0, 0, 0},
+		}
+	}
+
+	// If the path exists in the source info registry, return that object.
+	if loc, ok := si[strPath(path)]; ok {
+		return loc
+	}
+
+	// No dice; try the immediate parent path.
+	return si.findLocation(path[:len(path)-1])
```

Comment:
If not found, we should return `nil`, and let developers decide what to do.

Reply:
Again, this is intentional, moving up the chain to the closest possible match.

The premise of "return `nil` and let developers decide what to do" does not hold water -- at the point that this does the wrong thing, it is out of the rule author's hands.

Reply:
Please note that `SourceCodeInfo_Location` has comments information. If, for example, a rule wants to check the comments of a package, it is wrong to return the file's comments (which may be still `nil`). But take one step back, I am also Okay with what you do here -- I don't have strong objection, and feel free to make it as-it and resolve our conservation. Thank you!